### PR TITLE
Reset adam parameters on LR reset, better logging and MovieSampler fix

### PR DESCRIPTION
**Pull request by thomasZen**
- Re-initialize the Adam if the learning rate was changed by using a new class called OptimizerResetCallback
- Refactor MovieSampler and generate shifts from 0 to clip_chunk_size when not enforcing scene boundaries
- Better logging:
   - Log learning rate
   - Log training stats per epoch leading to smoother and easier comparable graphs
   - Log gradient magnitudes
- Change precision=16 to precision="16-mixed" as this is recommended by pytorch lightning
- Filter data by default based by removing low quality cells

# Results

## Low res
```
───────────────────────────────────────────────────────────────────────────────────────────────────────────────
       Test metric             DataLoader 0             DataLoader 1             DataLoader 2                                                                                                               
───────────────────────────────────────────────────────────────────────────────────────────────────────────────
    test_correlation        0.22427836060523987      0.2659996449947357       0.4651343822479248                                                                                                            
        test_loss            534.9968872070312        375.6442565917969        38.34758377075195                                                                                                            
───────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
With seed 42:
```
    test_correlation        0.20960338413715363      0.24668897688388824      0.4416184425354004                                                                               
        test_loss            535.2044067382812        377.0640869140625        37.99009704589844 
```

## High Res
With seed 42 (but one layer does not support deterministic mode):
```
────────────────────────────
       Test metric             DataLoader 0             DataLoader 1             DataLoader 2
───────────────────────────────────────────────────
    test_correlation        0.28346744179725647      0.33637067675590515       0.524970531463623
        test_loss            520.753662109375         364.3420715332031        38.09238052368164
────────────────────────────────────────────────────────────
```

### CODE REVIEW

File: configs/example_train_core_readout_low_res.yaml [hunk 1]
```diff
@@ -16,7 +20,7 @@ core_readout:
   gaussian_masks: True
   gaussian_mean_scale: 6.0
   gaussian_var_scale: 4.0
-  positive: True
+  positive: False
```
Comment (fededagos) [inline configs/example_train_core_readout_low_res.yaml]: Why? Lara's original config seems to be positive: https://github.com/open-retina/open-retina/blob/c2e60e2bf5189fd8a2491d01247621cc456be20b/openretina/hoefling_2024/configs.py#L8

File: openretina/models/core_readout.py [hunk 2]
```diff
@@ -373,6 +374,20 @@ def __init__(
...
 
+    def on_before_optimizer_step(self, optimizer):
+        # Compute the 2-norm for each layer
+        # If using mixed precision, the gradients are already unscaled here
+        core_norms = grad_norm(self.core, norm_type=2)
+        self.log_dict(core_norms, on_step=False, on_epoch=True)
+        readout_norms = grad_norm(self.readout, norm_type=2)
+        self.log_dict(readout_norms, on_step=False, on_epoch=True)
+
+    def get_last_lr(self) -> float:
```
Comment (fededagos) [inline openretina/models/core_readout.py]: There is an already implemented callback for lightning to do this: https://pytorch-lightning.readthedocs.io/en/1.2.10/api/pytorch_lightning.callbacks.lr_monitor.html Could we use it? Would avoid having to add this manually in the model.

It should also be more robust given that it is a lightning callback to use out of the box. You could maybe add this to the example config directly.

I think you could just copy these lines from one of the configs in the cell-type branch and it should work:
https://github.com/open-retina/open-retina/blob/c4d3f8527af77358055b399aa5ab9b4a0322d083/configs/minimal_session_agnostic_lightning.yaml#L52

**Review by fededagos — CHANGES_REQUESTED** [review]
Could you add the lr_monitor as a callback? Everything else looks good, though I also do not know why you changed readout_positive: False in Lara's config.

File: configs/example_train_core_readout_low_res.yaml (hunk 1 above)
Comment (thomasZen) [reply configs/example_train_core_readout_low_res.yaml]: Good point, I changed this back.

File: openretina/models/core_readout.py (hunk 2 above)
Comment (thomasZen) [reply openretina/models/core_readout.py]: Nice, I changed the config and got rid of this function

**Comment by thomasZen** [comment]
I also added the option to use a seed and then set the determistic option in pytorch lightning

**Review by fededagos — APPROVED** [review]
All good :)
